activerecord/lib/active_record/tasks/database_tasks.rb

```diff
@@ -192,9 +192,17 @@ def prepare_all
 
               seed = true
             end
+          end
+        end
 
-            migrate
-            dump_schema(db_config) if ActiveRecord.dump_schema_after_migration
+        each_current_environment(env) do |environment|
+          db_configs_with_versions(environment).sort.each do |version, db_configs|
+            db_configs.each do |db_config|
+              with_temporary_pool(db_config) do
+                migrate(version)
+                dump_schema(db_config) if ActiveRecord.dump_schema_after_migration
+              end
```

Comment:
👋 @fatkodima @eileencodes I think after this noticed our `db:prepare` took significantly longer. I haven't looked closely yet, but at a glance, this change makes it so that we're dumping the schema for every single migration (instead of one dump at the very end).

Reply:
Hello, yes, that could be the case if there are a bunch of pending migrations.
I have an idea how to fix it, so will open a PR.

Reply:
Nice! Thank you! We (Gusto) have hundreds of engineers working on a Rails monolith so we see about 10-20 migrations a day 😅

Reply:
It would be nice to keep track of which dbs actually received migrations so that we don't have to do a schema needlessly.

Reply:
Here's my attempt at fixing this on `main`. If it's good, I'll put up a backport to 7.1 as well.